### PR TITLE
fix(data): use Energon validation recipe for val split

### LINKED ISSUE #5626: [bug] Energon validation set is shuffled and never resets — validation loss is not reproducible
## What happens

The Energon val split is built with the **training** recipe —
`get_train_dataset(split_part="val")` in
`src/megatron/bridge/data/energon/base_energon_datamodule.py:178`.

So the validation stream is shuffled and infinitely repeating. The iterator is also created once
and never rebuilt, so each validation pass continues where the previous one stopped.

## Example

With `eval_interval=50`, `eval_iters=10`:

- validation at step 50 evaluates samples 1-320
- validation at step 100 evaluates samples 321-640
- ...

Every validation scores a **different, shuffled subset**. Val loss can move between steps purely
because the subset changed, and two runs of the same config produce different val curves.

Side effect of the infinite repeat: a val dataset smaller than `dp_size × num_workers` fails at
startup with `"Cannot repeat empty dataset indefinitely"` (`wrappers/repeat_dataset.py:57`).

## Expected

Validation evaluates the same samples every time it runs, so the curve reflects only the model.

## Suggested fix

Use Energon's val recipe for the val split — `get_val_dataset()` instead of
`get_train_dataset(split_part="val")`. It is unshuffled and finite, and its optional `limit`
wraps the data in `LimitDataset(reset_after_epoch=True)`, which rewinds after each pass — so
repeated validations replay the same samples.

This also removes the empty-dataset failure above, since nothing needs to repeat indefinitely.

Happy to open a PR if this looks right.

### PR DESCRIPTION
**Pull request by caohch-1**
# What does this PR do ?

Use Energon's validation dataset recipe for the validation split instead of constructing validation data through the training recipe.

# Changelog

- Route the training split through `get_train_dataset()` and the validation split through `get_val_dataset()`.
- Keep training-only shuffle configuration out of the validation path while preserving existing dataset-loading options.
- Add regression coverage for validation dataset recipe selection and update existing validation dataloader tests to use the validation builder.

# Validation

- A focused regression harness confirms that the pre-fix implementation selects `get_train_dataset()` for validation and the updated implementation selects `get_val_dataset()`.
- The same harness confirms that the training split continues to use `get_train_dataset()`.
- `git diff --check` passes.
- Python compilation checks pass for both changed files.

# GitHub Actions CI

This is an external contribution, so GitHub Actions CI will require approval from an NVIDIA developer.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed Contributor guidelines
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation? No documentation changes are needed for this fix.
- [ ] Does the PR affect components that are optional to install? No.
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

# Additional Information

Fixes #5626

### CODE REVIEW

File: src/megatron/bridge/data/energon/base_energon_datamodule.py [hunk 1]
```diff
@@ -171,24 +171,29 @@ def datasets_provider(self, worker_config, split: Literal["train", "val"] = "val
...
+                max_samples_per_sequence=self.max_samples_per_sequence,
+                **self.kwargs,
+            )
 
-        _dataset = get_train_dataset(
+        return get_val_dataset(
             self.path,
             batch_size=self.micro_batch_size,
-            task_encoder=task_encoder,
+            task_encoder=self.validation_task_encoder,
```
Comment (yaoyu-33) [inline src/megatron/bridge/data/energon/base_energon_datamodule.py:330]: validation_task_encoder will this always exist? or required

Comment (caohch-1) [reply src/megatron/bridge/data/energon/base_energon_datamodule.py:330]: Yes. `validation_task_encoder` is initialized in `__init__` as `validation_task_encoder or self.task_encoder`, so it falls back to the training task encoder when a separate validation encoder is not provided. The existing val path already used `self.validation_task_encoder`; this PR only changes the dataset builder from `get_train_dataset()` to `get_val_dataset()`.

If both encoders are `None` in direct construction, the value can still be None, but that is pre-existing behavior and is not introduced by this change.

**Comment by psycherun-kaiko** [comment]
@yaoyu-33 gentle ping :) 

This would unblock reliable val curves 📉
